File: backend/src/mod-filter/FileOperator.ts
import fs from "node:fs/promises";
import fsSync from "node:fs";
import path from "node:path";
import { logger } from "../utils/logger.js";

/**
 * 文件操作器 - 负责移动客户端模组
 */
export class FileOperator {
  private readonly movePath: string;

  constructor(movePath: string) {
    this.movePath = movePath;
  }

  /**
   * 移动客户端模组到回收目录
   */
  async moveClientSideMods(clientMods: string[]): Promise<{ success: number; error: number; skipped: number }> {
    if (!clientMods.length) {
      logger.info("No client-side mods to move");
      return { success: 0, error: 0, skipped: 0 };
    }

    // 确保目标目录存在
    const absoluteMovePath = path.isAbsolute(this.movePath) ? this.movePath : path.resolve(this.movePath);
    logger.debug("Target directory", { path: absoluteMovePath, exists: fsSync.existsSync(absoluteMovePath) });

    if (!fsSync.existsSync(absoluteMovePath)) {
      logger.debug("Creating target directory", { path: absoluteMovePath });
      await fs.mkdir(absoluteMovePath, { recursive: true });
    }

    let successCount = 0, errorCount = 0, skippedCount = 0;

    for (const sourcePath of clientMods) {
      try {
        // 转换为绝对路径
        const absoluteSourcePath = path.isAbsolute(sourcePath) ? sourcePath : path.resolve(sourcePath);

        logger.debug("Checking file", { originalPath: sourcePath, resolvedPath: absoluteSourcePath, cwd: process.cwd() });

        // 检查文件是否存在
        try {
          await fs.access(absoluteSourcePath);
        } catch (accessError) {
          logger.warn("File does not exist, skipping", { path: absoluteSourcePath, error: (accessError as Error).message });
          skippedCount++;
          continue;
        }

        // 提取文件名，处理特殊字符
        const filename = path.basename(absoluteSourcePath);
        const targetPath = path.join(absoluteMovePath, filename);

        logger.info("Moving file", { source: absoluteSourcePath, target: targetPath, filename: filename });

        // 使用 copyFile + unlink 代替 rename，避免 Windows 文件锁定问题
        await fs.copyFile(absoluteSourcePath, targetPath);
        await fs.unlink(absoluteSourcePath);

        successCount++;
      } catch (error: any) {
        logger.error("Failed to move file", { source: sourcePath, error: error.message, code: error.code });
        errorCount++;
      }
    }

    logger.info("File movement completed", { total: clientMods.length, success: successCount, error: errorCount, skipped: skippedCount });
    return { success: successCount, error: errorCount, skipped: skippedCount };
  }
}

File: core/dearth/fileoperator.go
package dearth

import (
	"os"
	"path/filepath"

	"deearthx/core/util"
)

// FileOperator moves identified client-side mods
type FileOperator struct {
	movePath string
}

// NewFileOperator creates a new FileOperator
func NewFileOperator(movePath string) *FileOperator {
	return &FileOperator{movePath: movePath}
}

// MoveResult contains the result of file movement
type MoveResult struct {
	Success int
	Error   int
	Skipped int
}

// MoveClientSideMods moves client-side mods to the rubbish directory
func (fo *FileOperator) MoveClientSideMods(clientMods []string) (*MoveResult, error) {
	if len(clientMods) == 0 {
		util.Logger.Debug("没有需要移动的客户端模组")
		return &MoveResult{}, nil
	}

	absoluteMovePath := fo.movePath
	if !filepath.IsAbs(absoluteMovePath) {
		absPath, err := filepath.Abs(absoluteMovePath)
		if err != nil {
			absoluteMovePath = fo.movePath
		} else {
			absoluteMovePath = absPath
		}
	}

	// Ensure target directory exists
	os.MkdirAll(absoluteMovePath, 0755)

	result := &MoveResult{}

	for _, sourcePath := range clientMods {
		absoluteSourcePath := sourcePath
		if !filepath.IsAbs(absoluteSourcePath) {
			absPath, err := filepath.Abs(sourcePath)
			if err != nil {
				absoluteSourcePath = sourcePath
			} else {
				absoluteSourcePath = absPath
			}
		}

		// Check if file exists
		if _, err := os.Stat(absoluteSourcePath); err != nil {
			util.Logger.Warn("文件不存在，跳过: " + absoluteSourcePath)
			result.Skipped++
			continue
		}

		filename := filepath.Base(absoluteSourcePath)
		targetPath := filepath.Join(absoluteMovePath, filename)

		util.Logger.Debug("正在移动文件: " + filename)

		// Copy file to target
		err := util.CopyFile(absoluteSourcePath, targetPath)
		if err != nil {
			util.Logger.Error("复制文件失败: " + err.Error())
			result.Error++
			continue
		}

		// Delete original file
		err = os.Remove(absoluteSourcePath)
		if err != nil {
			util.Logger.Error("删除原文件失败: " + err.Error())
			result.Error++
			continue
		}

		result.Success++
	}

	util.Logger.Info("文件移动完成",
		"总数", len(clientMods),
		"成功", result.Success,
		"失败", result.Error,
		"跳过", result.Skipped)

	return result, nil
}